File: C++ Version/src/OrderBook.cpp
#include "OrderBook.h"
#include <iostream>

OrderBook::OrderBook(std::string_view symbol)
    : m_symbol{ symbol }, m_bids{}, m_asks{}, m_trades{}
{}

void OrderBook::addOrder(const Order& order)
{
    if (order.getType() == OrderType::BUY)
    {
        m_bids.push(order);
    }
    else
    {
        m_asks.push(order);
    }
    matchOrders();
}

void OrderBook::matchOrders()
{
    while (!bids_or_ask_empty())
    {
        // Get the bid and ask orders
        Order bidOrder            = m_bids.top();
        Order askOrder            = m_asks.top();

        // Get the best bid and ask prices
        const double bestBidPrice = bidOrder.getPrice();
        const double bestAskPrice = askOrder.getPrice();

        if (bestBidPrice >= bestAskPrice)
        {
            const double tradePrice    = bestAskPrice;
            const int    tradeQuantity = std::min(
                bidOrder.getQuantity(), askOrder.getQuantity()
            );

            Trade trade{ bidOrder.getOrderID().data(),
                         askOrder.getOrderID().data(),
                         m_symbol,
                         tradePrice,
                         tradeQuantity };

            m_trades.push_back(trade);
            std::cout << "Trade Executed: " << trade.getRepresentation()
                      << '\n';

            m_asks.pop();
            m_bids.pop();

            int remainingBidQuantity = bidOrder.getQuantity() - tradeQuantity;
            int remainingAskQuantity = askOrder.getQuantity() - tradeQuantity;

            if (remainingBidQuantity > 0)
            {
                Order newBid = bidOrder;
                newBid.setQuantity(remainingBidQuantity);
                m_bids.push(newBid);
            }

            if (remainingAskQuantity > 0)
            {
                Order newAsk = askOrder;
                newAsk.setQuantity(remainingAskQuantity);
                m_asks.push(newAsk);
            }
        }
        else
        {
            break;
        }
    }
}

void OrderBook::displayOrderBook() const
{
    std::cout << "Order Book for " << m_symbol << '\n';
    std::cout << "Buy Orders:\n";
    std::priority_queue<Order, std::vector<Order>, BuyOrderComparator>
        bidsCopy = m_bids;
    while (!bidsCopy.empty())
    {
        std::cout << bidsCopy.top().getRepresentation() << '\n';
        bidsCopy.pop();
    }

    std::cout << "Sell Orders:\n";
    std::priority_queue<Order, std::vector<Order>, SellOrderComparator>
        asksCopy = m_asks;
    while (!asksCopy.empty())
    {
        std::cout << asksCopy.top().getRepresentation() << '\n';
        asksCopy.pop();
    }
}

void OrderBook::displayTrades() const
{
    std::cout << "Trades for " << m_symbol << '\n';
    for (const Trade& trade : m_trades)
    {
        std::cout << trade.getRepresentation() << '\n';
    }
}

File: Python Version/order_book.py
import heapq
from order import Order
from order_type import OrderType
from trade import Trade
from order_book_enums import OrderBookSide


class OrderBook:
    def __init__(self, symbol: str):
        self.symbol = symbol
        self.bids = []  # Highest price first (max heap)
        self.asks = []  # Lowest price first (min heap)
        self.trades = []

    def add_order(self, order: Order) -> None:
        if order.order_type == OrderType.BUY.value:
            order.price = -order.price
            heapq.heappush(self.bids, (order))
        else:
            heapq.heappush(self.asks, (order))
        self.match_orders()

    def match_orders(self) -> None:
        while self.bids and self.asks:
            bid_order = self.bids[0]
            ask_order = self.asks[0]

            best_bid = -bid_order.price
            best_ask = ask_order.price

            if best_bid >= best_ask:
                trade_price = best_ask
                trade_quantity = min(bid_order.quantity, ask_order.quantity)

                trade = Trade(
                    bid_order.order_id,
                    ask_order.order_id,
                    self.symbol,
                    trade_quantity,
                    trade_price,
                )
                self.trades.append(trade)
                print(f"Trade executed: {trade}")

                bid_order.quantity -= trade_quantity
                ask_order.quantity -= trade_quantity

                heapq.heappop(self.bids)
                if bid_order.quantity > 0:
                    heapq.heappush(self.bids, (bid_order))

                heapq.heappop(self.asks)
                if ask_order.quantity > 0:
                    heapq.heappush(self.asks, (ask_order))
            else:
                break

    def get_trades(self) -> list:
        return self.trades

    def get_order_book(self) -> dict:
        return {OrderBookSide.BID.value: self.bids, OrderBookSide.ASK.value: self.asks}
